frame/assets/src/functions.rs

```diff
@@ -652,71 +661,125 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 				sufficients: 0,
 				approvals: 0,
 				is_frozen: false,
+				status: AssetStatus::Live,
 			},
 		);
 		Self::deposit_event(Event::ForceCreated { asset_id: id, owner });
 		Ok(())
 	}
 
-	/// Destroy an existing asset.
-	///
-	/// * `id`: The asset you want to destroy.
-	/// * `witness`: Witness data needed about the current state of the asset, used to confirm
-	///   complexity of the operation.
-	/// * `maybe_check_owner`: An optional check before destroying the asset, if the provided
-	///   account is the owner of that asset. Can be used for authorization checks.
-	pub(super) fn do_destroy(
+	/// Start the process of destroying an asset, by setting the asset status to Destroying, and
+	/// emitting the DestructionStarted event.
+	pub(super) fn do_start_destroy(
 		id: T::AssetId,
-		witness: DestroyWitness,
 		maybe_check_owner: Option<T::AccountId>,
-	) -> Result<DestroyWitness, DispatchError> {
-		let mut dead_accounts: Vec<T::AccountId> = vec![];
-
-		let result_witness: DestroyWitness = Asset::<T, I>::try_mutate_exists(
-			id,
-			|maybe_details| -> Result<DestroyWitness, DispatchError> {
-				let mut details = maybe_details.take().ok_or(Error::<T, I>::Unknown)?;
+	) -> DispatchResult {
+		let _ =
+			Asset::<T, I>::try_mutate_exists(id, |maybe_details| -> Result<(), DispatchError> {
+				let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
 				if let Some(check_owner) = maybe_check_owner {
 					ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
 				}
-				ensure!(details.accounts <= witness.accounts, Error::<T, I>::BadWitness);
-				ensure!(details.sufficients <= witness.sufficients, Error::<T, I>::BadWitness);
-				ensure!(details.approvals <= witness.approvals, Error::<T, I>::BadWitness);
+				ensure!(details.is_frozen, Error::<T, I>::NotFrozen);
+				details.status = AssetStatus::Destroying;
+
+				Self::deposit_event(Event::DestructionStarted { asset_id: id });
+				Ok(())
+			})?;
+		Ok(())
+	}
+
+	/// Destroy accounts associated with a given asset up to the max (T::RemoveItemsLimit).
+	///
+	/// Each call emits the `Event::DestroyedAccounts` event.
+	pub(super) fn do_destroy_accounts(id: T::AssetId) -> Result<u32, DispatchError> {
+		let mut dead_accounts: Vec<T::AccountId> = vec![];
+		let mut removed_accounts = 0;
```

Comment:
```suggestion
```
You can just use `dead_accounts.len()`.

Reply:
@ggwpez I could change it, it's definitely more elegant that way, but it would make the code slower. 
```rust
if removed_accounts >= T::RemoveItemsLimit::get() {
```

I'm comparing removed_accounts in a loop, `dead_accounts.len()` is O(n) to get the length, so the logic would be n*O(n). It might be expensive for larger assets to calculate the length in each run of the loop. 
What do you think? Still worth doing?

Reply:
Vec::len` has constant runtime, not linear.

Reply:
Oh, I had no idea. Thanks for clarifying. I've made the update.